libnvme/libnvme.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (C) 2017 Red Hat, Inc.
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Gris Ge <fge@redhat.com>
+ */
+
+#include <libnvme/libnvme.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+#include "utils.h"
+#include "sysfs.h"
+#include "ctrl.h"
+
+struct _num_str_conv {
+	const int num;
+	const char *str;
+};
+
+static const struct _num_str_conv _NVME_RC_MSG_CONV[] = {
+	{NVME_OK, "OK"},
+	{NVME_ERR_NO_MEMORY, "Out of memory"},
+	{NVME_ERR_BUG, "BUG of libnvme library"},
+	{NVME_ERR_NO_SUPPORT, "Not supported"},
+	{NVME_ERR_PERMISSION_DENY, "Permission deny"},
+};
+
+int nvme_ctrls_get(struct nvme_ctrl ***cnts, uint32_t *cnt_count,
+		   const char **err_msg)
+{
+	int rc = NVME_OK;
+	uint32_t i = 0;
+	const char **nvme_ctrl_dev_paths = NULL;
+	uint32_t count = 0;
+	char err_msg_buff[_NVME_ERR_MSG_BUFF_LEN];
+	char *tmp_err_msg = NULL;
+
+	assert(cnts != NULL);
+	assert(cnt_count != NULL);
+
+	_nvme_err_msg_clear(err_msg_buff);
+
+	*cnts = NULL;
+	*cnt_count = 0;
+	if (err_msg != NULL)
+		*err_msg = NULL;
+
+	_good(_sysfs_nvme_ctrl_dev_paths_get(err_msg_buff,
+					     &nvme_ctrl_dev_paths,
+					     &count),
+	      rc, out);
+
+	if (count == 0)
+		goto out;
+
+	*cnts = (struct nvme_ctrl **)
+		calloc(count, sizeof(struct nvme_ctrl *));
+	_alloc_null_check(err_msg_buff, *cnts, rc, out);
+	*cnt_count = count;
+
+	for (i = 0; i < count; ++i)
+		_good(_nvme_ctrl_new(nvme_ctrl_dev_paths[i], &((*cnts)[i]),
+				     err_msg_buff),
+		      rc, out);
+
+out:
+	_sysfs_nvme_ctrl_dev_paths_free(nvme_ctrl_dev_paths, count);
+	if (rc != NVME_OK) {
+		nvme_ctrls_free(*cnts, *cnt_count);
+		*cnts = NULL;
+		*cnt_count = 0;
+		if (err_msg != NULL) {
+			tmp_err_msg = (char *)
+				malloc(sizeof(char *) *
+				       _NVME_ERR_MSG_BUFF_LEN);
+			if (tmp_err_msg != NULL) {
+				memcpy(tmp_err_msg, err_msg_buff,
+				       _NVME_ERR_MSG_BUFF_LEN);
+				*err_msg = tmp_err_msg;
+			}
+		}
+	}
+	return rc;
+}
+
+void nvme_ctrls_free(struct nvme_ctrl **cnts, uint32_t cnt_count)
+{
+	uint32_t i = 0;
+
+	if ((cnts == NULL) || (cnt_count == 0))
+		return;
+
+	for (i = 0; i < cnt_count; ++i)
+		nvme_ctrl_free(cnts[i]);
+	free(cnts);
+}
+
+void nvme_ctrl_pro_raw_get(struct nvme_ctrl *cnt, const char *pro_name,
+			   uint8_t **buff, size_t *size, const char **err_msg)
```

Comment:
This function doesn't appear to be used.., it's signature and it's implementation don't seem to correlate too?